Control/Concurrent/Async.hs

```diff
@@ -24,74 +24,134 @@
 -- @IO@ action that will return a value of type @a@, or die with an
--- exception.  An @Async@ corresponds to a thread, and its 'ThreadId'
--- can be obtained with 'asyncThreadId', although that should rarely
--- be necessary.
+-- exception.  An 'Async' is a wrapper around a low-level 'forkIO' thread.
+--
+-- The fundamental function to spawn threads with the high-level API is
+-- 'withAsync'.
 --
 -- For example, to fetch two web pages at the same time, we could do
 -- this (assuming a suitable @getURL@ function):
 --
--- >    do a1 <- async (getURL url1)
--- >       a2 <- async (getURL url2)
--- >       page1 <- wait a1
--- >       page2 <- wait a2
--- >       ...
+-- > withAsync (getURL url1) $ \a1 -> do
+-- >   withAsync (getURL url2) $ \a2 -> do
+-- >     page1 <- wait a1
+-- >     page2 <- wait a2
+-- >     ...
+--
+-- where 'withAsync' starts the operation in a separate thread, and
+-- 'wait' waits for and returns the result.
+--
+-- * If the operation throws an exception, then that exception is re-thrown
+--    by 'wait'. This ensures property (1): No exception is swallowed.
+-- * If an exception bubbles up through a 'withAsync', then the 'Async'
+--   it spawned its 'cancel'ed. This ensures property (2): No thread is leaked.
+--
+-- Often we do not care to work manually with 'Async' handles like
+-- @a1@ and @a2@. Instead, we want to express high-level objectives like
+-- performing two or more tasks concurrently, and waiting for one or all
+-- of them to finish.
+--
+-- For example, the pattern of performing two IO actions concurrently and
+-- waiting for both their results is packaged up in a combinator 'concurrently',
+-- so we can further shorten the above example to:
+--
+-- > (page1, page2) <- concurrently (getURL url1) (getURL url2)
+-- > ...
 --
--- where 'async' starts the operation in a separate thread, and
--- 'wait' waits for and returns the result.  If the operation
--- throws an exception, then that exception is re-thrown by
--- 'wait'.  This is one of the ways in which this library
--- provides some additional safety: it is harder to accidentally
--- forget about exceptions thrown in child threads.
+-- The section __/High-level utilities/__ covers the most
+-- common high-level objectives, including:
 --
--- A slight improvement over the previous example is this:
+-- * Waiting for 2 results ('concurrently').
+-- * Waiting for /N/ many results ('mapConcurrently' / 'forConcurrently').
+-- * Waiting for the first of 2 results ('race').
+-- * Waiting for arbitrary nestings of "all of /N/" and "the first of /N/"
+--   results with the 'Concurrently' newtype and its 'Applicative' and
+--   'Alternative' instances.
 --
--- >       withAsync (getURL url1) $ \a1 -> do
--- >       withAsync (getURL url2) $ \a2 -> do
--- >       page1 <- wait a1
--- >       page2 <- wait a2
--- >       ...
+-- Click here to scroll to that section:
+-- "Control.Concurrent.Async#high-level-utilities" (and help implement
```

Comment:
is this pointer to Haddock relevant here?

Reply:
The problem is that haddock renders the link as just `Control.Concurrent.Async`, which is very confusing beacuse it suggests that it links to the full module, not a section, see screenshot:

![image](https://user-images.githubusercontent.com/399535/73941585-d36ff400-48ed-11ea-801c-ba9ef050199e.png)

I was a bit on the fence whether I should put this call to action to fix it in; I kept it in the first draft so you can see it and decide whether you think it's worth it or whether we should keep async's docs free from petitions to help with tooling :D

Reply:
I think "Click here to scroll to that section" is enough of a hint (perhaps "jump" rather than "scroll" though), let's remove the link to the Haddock issue.

Reply:
OK, done.